[K1mcheee] iP

K1mcheee opened the pull request:
# Mocha bot

## Why Mocha

The name was inspired from javascript's old name (mocha) before it got renamed (_a few times_) to javascript. Fun fact, the rename was partially to [leverage on the popularity of java](https://medium.com/@salimian/the-story-behind-the-name-javascript-15e879c42760)!

Oh, you're wondering _why use my bot_ ?

Mocha frees you from the pain of forgetting tasks you need to do. It's:

* text-based
* easy to learn
* fast to use

All you need to do is,

1. Download it from [here]()
1. double-click it
1. add your tasks
1. let it **help you** remember

best part of it? It is completely **FREE**

> _The best things in life are free_

## Features

- [x] Managing tasks
- [x] Tracking deadlines
- [ ] Reminders (Coming soon, keep a lookout :eyes:)

If you are a Java programmer, you can use it to practice Java too ~~and remember how javascript got its name~~  
Here's an example of the `main` class:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

K1mcheee committed:
Renamed Duke to Mocha

K1mcheee committed:
Add greet function

K1mcheee committed:
Add echo functionality; fix access modifier of BR

K1mcheee committed:
Add storing input functionality

K1mcheee committed:
Added printing list functionality

K1mcheee committed:
Added class for Tasks

K1mcheee committed:
Added mark, unmark and toString functions to Task

K1mcheee committed:
Integrated Task processing into Mocha, fixed printing of tasks

K1mcheee committed:
Add input validation and fixed printing

K1mcheee committed:
Abstracted task status check into method and fixed error type and printing

K1mcheee committed:
Added todo event and deadline classes

K1mcheee committed:
Integrated todo to Mocha, fixed printing of tasks

K1mcheee committed:
Integrated deadline tasks, fix spacing issue for printing of tasks

K1mcheee committed:
Integrated events into Mocha

K1mcheee committed:
Added javadocs to tasks and its subclasses

K1mcheee committed:
Add testcases and modify runtest script

K1mcheee committed:
fix printing problem

K1mcheee committed:
Add a custom exception class

K1mcheee committed:
Add exception handling with custom messages

K1mcheee committed:
Add delete function, modify exception message

K1mcheee committed:
Add javadocs

K1mcheee committed:
Edit code format

K1mcheee committed:
Create new TaskFile for storing task list

K1mcheee committed:
Add functionality to save list to drive

K1mcheee committed:
Merge branch 'branch-Level-7'

K1mcheee committed:
Fix some printing issues for saved data

K1mcheee committed:
Fix more save issues

K1mcheee committed:
Apply changes to Event and Deadline tasks

K1mcheee committed:
More fixes for level 7

K1mcheee committed:
Add Date and Time functionality for Deadline tasks

K1mcheee committed:
Extend date and time implementation to events

K1mcheee committed:
Fix bugs and implement printing of due tasks

K1mcheee committed:
Merge branch 'branch-level-8'

K1mcheee committed:
Add Ui class

K1mcheee committed:
Create TaskList class

K1mcheee committed:
Create Command classes

K1mcheee committed:
Add Parser class and create more command classes

K1mcheee committed:
Add remaining command classes and fix printing issues

K1mcheee committed:
Fix parser class and implement parsing and commands

K1mcheee committed:
Merge branch 'branch-A-OOP'

K1mcheee committed:
Refactor into packages

K1mcheee committed:
Merge branch 'branch-A-Packages'

K1mcheee committed:
Merge branch 'add-gradle-support'

K1mcheee committed:
Add DeadlineTest, fix issues

K1mcheee committed:
Add ParserTest

K1mcheee committed:
Merge branch 'branch-A-JUnit'

K1mcheee committed:
Fix errors

K1mcheee committed:
Add javadocs

K1mcheee committed:
Fix style

K1mcheee committed:
Create Find command

K1mcheee committed:
Merge branch 'branch-A-JavaDoc'

K1mcheee committed:
Merge branch 'branch-A-CodingStandard'

K1mcheee committed:
Merge branch 'branch-Level-9'

K1mcheee committed:
Fix Find command and some formatting

nicolaskjh commented on src/main/java/mocha/Mocha.java:
```diff
@@ -0,0 +1,48 @@
+ * @author K1mcheee
+ */
+public class Mocha {
+    private static boolean isRunning = true;
```
Good job following the naming convention for boolean attributes!

nicolaskjh commented on src/main/java/mocha/Parser.java:
```diff
@@ -0,0 +1,105 @@
+                }
+            } else {
+                // switch case to respond to input
+                switch (tmp) {
```
Good job following the indentation rules for a switch case block

nicolaskjh commented on src/main/java/mocha/Parser.java:
```diff
@@ -0,0 +1,105 @@
+                    if (tmp.equals("delete")) {
+                        c = new DeleteCommand(idx);
+                    }
+                }
```
This block of code seems highly nested, perhaps an else if for each command type could reduce the need for this much indentations?

nicolaskjh commented on src/main/java/mocha/Parser.java:
```diff
@@ -0,0 +1,105 @@
+        // parse input
+        String[] split = input.split(" ");
+        String[] date = input.split("/");
+        String tmp = split[0].toLowerCase();
```
Perhaps a more descriptive variable name like commandType could be used?

nicolaskjh commented on src/main/java/mocha/Parser.java:
```diff
@@ -0,0 +1,105 @@
+        }  else if (input.toLowerCase().equals("due")) {
+            // check for command to print due tasks
+            c = new DueCommand();
+        }   else if (input.toLowerCase().equals("list")) {
```
Be careful of the number of spaces after the curly brace before the else if!

nicolaskjh commented on src/main/java/mocha/TaskList.java:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @return size of List.
+     */
+    public int size() {
```
Maybe a more descriptive method name that is a verb like getSize() would be better!

nicolaskjh commented on src/main/java/mocha/Ui.java:
```diff
@@ -0,0 +1,89 @@
+
+public class Ui {
+    /**Breakline to separate texts*/
+    private static final String BR = "____________________________________________________________";
```
Good job on following the naming conventions for constants!

nicolaskjh commented on src/main/java/mocha/TaskFile.java:
```diff
@@ -0,0 +1,103 @@
+import mocha.task.Task;
+import mocha.task.Todo;
+
+import java.io.*;
```
Remember to explicitly list out imported classes!

nicolaskjh commented on src/main/java/mocha/TaskFile.java:
```diff
@@ -0,0 +1,103 @@
+package mocha;
+
+import mocha.task.Deadline;
+import mocha.task.Event;
+import mocha.task.Task;
+import mocha.task.Todo;
+
+import java.io.*;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
```
Good job on keeping the import ordering consistent and explicitly listing out the rest of the imported classes!

nicolaskjh reviewed (commented):
Good job so far! Keep up the good work! :)

thegrimbee commented on src/main/java/mocha/TaskList.java:
```diff
@@ -0,0 +1,62 @@
+
+public class TaskList {
+    /**List of tasks*/
+    private List<Task> commands; //tasklist
```
It's great that you are using plural form for collections, but maybe a better name would be tasks since this is unrelated to the command classes!

thegrimbee commented on data/mocha.txt:
```diff
@@ -0,0 +1,5 @@
+0 event sleep /from 2025-01-30 /to 2025-02-01
+1 todo eat
+1 todo sleep
+0 deadline sleep /by 2025-01-30
```
It might be better to consider including this file in gitignore, so that when users download the project, they can have an empty tasklist

thegrimbee commented on src/main/java/mocha/task/Deadline.java:
```diff
@@ -0,0 +1,132 @@
+        this.deadlineTime = null;
+    }
+
+    public Deadline(String name, LocalDateTime deadlineTime) {
```
Great use of polymorphism!

thegrimbee commented on src/main/java/mocha/task/Event.java:
```diff
@@ -0,0 +1,192 @@
+/**
+ * Encapsulates an Event task.
+ *
+ * @author K1mcheee
```
I like how you include @author in the javadocs, very professional!

thegrimbee commented on src/main/java/mocha/Parser.java:
```diff
@@ -0,0 +1,105 @@
+            c = new ListCommand();
+        } else {
+            // check for keywords mark and unmark
+            if (tmp.equals("mark") || tmp.equals("unmark")
```
Just a possible alternative: instead of setting the command names in parser, we can add a getKeyword() method to each command to make use of abstraction! E.g. getKeyword() for MarkCommand would return "mark". That way, in parser, we just need to check if tmp equals the command keyword.

thegrimbee commented on src/main/java/mocha/command/ByeCommand.java:
```diff
@@ -0,0 +1,27 @@
+     * @param tasks List of current tasks.
+     * @param ui Interface for users to interact with.
+     * @param storage Stores updates and changes to drive.
+     */
```
I really like how clean and concise all the javadocs are!

thegrimbee reviewed (commented):
LGTM, code is very clean! Just some minor suggestions and nitpicks

K1mcheee committed:
Add Main, MainWindow and DialogBox classes

K1mcheee committed:
Add more classes for javaFX

K1mcheee committed:
configure for javafx

K1mcheee committed:
Add functionality to GUI

K1mcheee committed:
Merge branch 'branch-Level-10'

K1mcheee committed:
Add assertions

K1mcheee committed:
Fix code style, implement SLAP

K1mcheee committed:
Implement slap and fix code quality issues

K1mcheee committed:
Fix other code quality and bugs

K1mcheee committed:
Fix more formatting

K1mcheee committed:
Merge pull request #1 from K1mcheee/branch-A-CodeQuality

Branch a code quality

K1mcheee committed:
Merge branch 'branch-A-Assertions' into master

K1mcheee committed:
Merge pull request #4 from K1mcheee/master

Merge master to branch

K1mcheee committed:
Revert "Add assertions"

K1mcheee committed:
Merge pull request #5 from K1mcheee/revert-2-branch-A-Assertions

Revert "Add assertions"

K1mcheee committed:
Merge pull request #6 from K1mcheee/branch-A-Assertions

Merge pull request #4 from K1mcheee/master

K1mcheee committed:
App has no utility to add a tag to tasks

There is a need to add tags so that future utilities like sort and filter can work better

Add a TagCommand and implement tag function on tasks

User can include a tag after adding the tasks, so that they can group them into relevant groups.

K1mcheee committed:
Tag does not get saved

Important to save the tags as user would want their tags to remain after closing the app

Update TaskFile to save the tags and fix some additional bugs. Add function to search tasks using tags.

This allows for users to retain the tags saved and search for tasks based on tags.

K1mcheee committed:
Add untag function and fix loading and saving errors from tag

K1mcheee committed:
Merge branch 'branch-BCD-Extension'

K1mcheee committed:
Missed out tags for week 5 commits

K1mcheee committed:
Add some personality to response

K1mcheee committed:
Add error handling for to date before from date

K1mcheee committed:
Modify split to handle case of extra whitespace

K1mcheee committed:
Merge branch 'branch-A-MoreErrorHandling'

K1mcheee committed:
Delete unused method

K1mcheee committed:
Fix parsing of find command and parsertest

K1mcheee committed:
Handle error for tagging

K1mcheee committed:
Add user guide

K1mcheee committed:
Fix user guide

K1mcheee committed:
Fix code blocks for README

K1mcheee committed:
Test fix for README code format

K1mcheee committed:
Create _config.yml

Add theme to site

K1mcheee committed:
Add theme for site

K1mcheee committed:
Merge branch 'master' of https://github.com/K1mcheee/ip

K1mcheee committed:
Update theme

K1mcheee committed:
Fix code formatting problem

K1mcheee committed:
Fix header links

K1mcheee committed:
Add more command formats to README

K1mcheee committed:
Remove unused file and fix other format issues in code

K1mcheee committed:
Fix errors that occur with invalid user input

K1mcheee committed:
Add input verification to prevent errors

K1mcheee committed:
Add documentation for due command
